### ISSUE TITLE
Enhancement: ship data consistency proofs; block query API

### ISSUE THREAD

**Issue by cc32d9** [eosnetworkfoundation/mandel#22]
Currently there's no way for a state history reader to verify that the data is consistent and not corrupted. the protocol needs to be enhanced to provide that, as many applications rely on the data.

Proposal: 

in each data frame of trace and state history that corresponds to one blockchain block, add two hashes:

1. sha256 checksum of the frame content
2. sha256 checksum of the previous block's and current block's checksums, concatenated.

The portable snapshot format would have to be different for ship and non-ship nodes: the one for ship nodes would have to include the checksums of latest ship frames.

Also these checksums will need to be compared across the nodes, so a new HTTP API is needed that delivers the state history data (compressed) in raw form for a specific block. Such an API would also be useful for history solutions which need to decode a specific random block from the past.

**Comment by swatanabe** [comment]
I don't think this will work.  SHiP doesn't have to start from genesis.

**Comment by cc32d9** [comment]
well, in this scenario there has to be at least one copy of ship that stems from genesis. It is a drawback, but the benefit would be a better consistency control.

probably instead of the previous frame hash, we should use a hash of (block ID, the current frame hash) tuple. This will let us check the consistency quickly without having to decode the contents.

**Comment by swatanabe** [comment]
Another problem is that SHiP data changes with each nodeos release

**Comment by cc32d9** [comment]
yes, but we need to approach the problem of trusting the data sources.

probably the ship provider needs to sign its output with a well-known key

**Comment by edrotthoff** [comment]
FIO is definitely interested in the process and actions that affect this issue, we are currently in discovery phase for an off chain state database and this is one of the key requirements, that the data be easily/quickly verifiable with the blockchain WRT its authenticity.
